src/local-development/clone.ts

```diff
@@ -0,0 +1,128 @@
+import * as vscode from 'vscode';
+import * as path from 'path';
+import { existsSync } from 'fs';
+import { TextEncoder } from 'util';
+import pick from 'lodash.pick';
+import { getCurrentWorkspace } from '../services/workspace';
+import { getCurrentEnvironment } from '../providers/configuration';
+import { LocalAppOriginWithSecret, MakecomappJson } from './types/makecomapp.types';
+import App from '../tree/App';
+import { askForAppDirToClone } from './ask-local-dir';
+import { APIKEY_DIRNAME, MAKECOMAPP_FILENAME } from './consts';
+import { generalCodesDefinition } from '../services/component-code-def';
+import { downloadSource } from './code-deploy-download';
+import { generateDefaultLocalFilename } from './local-file-paths';
+import { catchError } from '../error-handling';
+import { pullNewComponents } from './pull';
+import { storeSecret } from './secrets-storage';
+import { withProgressDialog } from '../utils/vscode-progress-dialog';
+import { entries } from '../utils/typed-object';
+
+export function registerCommands(): void {
+	vscode.commands.registerCommand(
+		'apps-sdk.local-dev.clone-to-workspace',
+		catchError(
+			'Clone app to workspace',
+			cloneAppToWorkspace,
+		),
+	);
+}
+
+/**
+ * Clones whole app from Make cloud to the local repository
+ */
+async function cloneAppToWorkspace(context: App): Promise<void> {
+	const workspaceRoot = getCurrentWorkspace().uri;
+	const apikeyDir = vscode.Uri.joinPath(workspaceRoot, APIKEY_DIRNAME);
+
+	const localAppRootdir = await askForAppDirToClone();
+	if (!localAppRootdir) {
+		return;
+	}
+
+	const environment = getCurrentEnvironment();
+
+	// makecomapp.json
+	const makeappJsonPath = vscode.Uri.joinPath(localAppRootdir, MAKECOMAPP_FILENAME);
+	// If manifest exists, cancel this task.
+	if (existsSync(makeappJsonPath.fsPath)) {
+		throw new Error(MAKECOMAPP_FILENAME + ' already exists in the workspace. Clone cancelled.');
+	}
+
+	const apikeyFileUri = await storeSecret('apikey', environment.apikey);
+
+	const origin: LocalAppOriginWithSecret = {
+		label: 'Origin',
+		baseUrl: 'https://' + environment.url,
+		appId: context.name,
+		appVersion: context.version,
+		apikeyFile: path.posix.relative(localAppRootdir.path, apikeyFileUri.path),
```

Comment:
Exactly, @mikulassychra, the key can be different for each app. Additionally, one app can have multiple API keys in case it has multiple origins, where to deploy itself.